services/hanke-service/src/integrationTest/kotlin/fi/hel/haitaton/hanke/hakemus/HakemusServiceITest.kt

```diff
@@ -182,4 +209,236 @@ class HakemusServiceITest(
+            }
+
+            exception.all {
+                hasClass(ApplicationGeometryNotInsideHankeException::class)
+                messageContains("id=${application.id}")
+                messageContains("hankeId=${hanke.id}")
+                messageContains("geometry=${notInHankeArea.geometry.toJsonString()}")
+            }
+        }
+
+        @Test
+        fun `saves updated data and creates an audit log`() {
+            val hanke = hankeFactory.builder(USERNAME).withHankealue().saveEntity()
+            val application =
+                createHakemusWithHakija(hanke) {
+                    applicationData =
+                        (applicationData as CableReportApplicationData).copy(
+                            workDescription = "Old work description"
+                        )
+                }
+            val yhteystieto = hakemusyhteystietoRepository.findAll().first()
+            val kayttaja =
+                hankekayttajaRepository
+                    .findByHankeIdAndSahkopostiIn(hanke.id, listOf(KAYTTAJA_INPUT_HAKIJA.email))
+                    .single()
+            val newKayttaja = hankeKayttajaFactory.saveUser(hanke.id)
+            val originalAuditLogSize = auditLogRepository.findByType(ObjectType.APPLICATION).size
+            val request =
+                HakemusUpdateRequestFactory
+                    .createJohtoselvityshakemusUpdateRequestFromApplicationEntity(application)
+                    .withCustomerWithContactsRequest(
+                        CustomerType.COMPANY,
+                        yhteystieto.id,
+                        kayttaja.id,
+                        newKayttaja.id
+                    )
+                    .withWorkDescription("New work description")
+
+            val updatedHakemus = hakemusService.updateHakemus(application.id!!, request, USERNAME)
+
+            assertThat(updatedHakemus.applicationData as JohtoselvitysHakemusDataResponse).all {
+                prop(JohtoselvitysHakemusDataResponse::workDescription)
+                    .isEqualTo("New work description")
+                prop(JohtoselvitysHakemusDataResponse::customerWithContacts)
+                    .isNotNull()
+                    .prop(CustomerWithContactsResponse::contacts)
+                    .transform { it.map { contact -> contact.hankekayttajaId } }
+                    .containsExactlyInAnyOrder(kayttaja.id, newKayttaja.id)
+            }
+            val applicationLogs = auditLogRepository.findByType(ObjectType.APPLICATION)
+            assertThat(applicationLogs).hasSize(originalAuditLogSize + 1)
+        }
+
+        private fun createHakemus(
+            hankeEntity: HankeEntity = hankeFactory.builder(USERNAME).withHankealue().saveEntity(),
+            f: ApplicationEntity.() -> Unit = {}
+        ): ApplicationEntity {
+            return hakemusFactory.builder(USERNAME, hankeEntity).withModification(f).save()
+        }
```

Comment:
I'm not a big fan of adding private factory methods to test classes. Developing the shared factory classes seems more productive in the long run.

Creating a hakemus could be easier by adding a default value for the hankeEntity parameter or the `builder()` method.
```
    fun builder(
        userId: String,
        hankeEntity: HankeEntity = hankeFactory.builder(userId).withHankealue().saveEntity()
    ): HakemusBuilder {
```

Then the simplest way to create a hakemus would be easy enough not to require a helper method.
```
    val application = hakemusFactory.builder(USERNAME).save()
```

The function parameter here is used only once to set the alluId of an application, which could be done with the existing (though private) `.withStatus()` method:
```
    val application = hakemusFactory.builder(USERNAME).withStatus(alluId = 21).save()
```

Reply:
I refactored the creation of ApplicationEntity.